discojs/src/privacy.ts

```
@@ -28,12 +102,29 @@ export function addNoise(
  **/
 export async function clipNorm(
   weights: WeightsContainer,
-  radius: number,
+  radius: number | number[],
 ): Promise<WeightsContainer> {
-  if (radius <= 0) throw new Error("invalid radius");
+  if (typeof radius === "number" && radius <= 0) 
+    throw new Error("invalid radius");
+
+  if (typeof radius === "number"){
+    const norm = await frobeniusNorm(weights);
+    const scaling = Math.max(1, norm / radius);
+
+    return weights.map((w) => w.div(scaling));
+  }else{
+    /** Apply different clipping radius to each layer in the WeightsContainer */
+    const layers = weights.weights;
+
+    const clipped = await Promise.all(
+      layers.map(async (l, i) => {
+        const norm = await layerWiseFrobeniusNorm(l);
+        const scaling = Math.max(1, norm / radius[i]);
```

we don't actually check for `radius[i]` is valid as done for the `number` path, better to throw than generating invalid value (`x / 0`  w/ `x > 0` gives `Infinity`)